File: IPAddress/src/inet.ipaddr/inet/ipaddr/format/validate/ParsedHost.java
/*
 * Copyright 2016-2018 Sean C Foley
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *     or at
 *     https://github.com/seancfoley/IPAddress/blob/master/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package inet.ipaddr.format.validate;

import java.io.Serializable;

import inet.ipaddr.AddressStringException;
import inet.ipaddr.HostName;
import inet.ipaddr.IPAddress;
import inet.ipaddr.IPAddress.IPVersion;
import inet.ipaddr.IPAddressNetwork;
import inet.ipaddr.IPAddressSection;
import inet.ipaddr.IPAddressString;

/**
 * The result of parsing a valid host name.
 * 
 * @author sfoley
 *
 */
public class ParsedHost implements Serializable {

	private static final long serialVersionUID = 4L;

	private static final EmbeddedAddress NO_EMBEDDED_ADDRESS = new EmbeddedAddress();
	static final ParsedHostIdentifierStringQualifier NO_QUALIFIER = new ParsedHostIdentifierStringQualifier();
	
	private String normalizedLabels[];
	private int separatorIndices[];
	private boolean normalizedFlags[];
	
	private final ParsedHostIdentifierStringQualifier labelsQualifier;
	private String service;
	
	private EmbeddedAddress embeddedAddress;
	
	String host;
	private final String originalStr;
	
	public ParsedHost(String originalStr, IPAddressProvider valueProvider) {
		this(originalStr, null, null, NO_QUALIFIER, new EmbeddedAddress());
		embeddedAddress.addressProvider = valueProvider;
	}
	
	public ParsedHost(String originalStr, IPAddressProvider valueProvider, ParsedHostIdentifierStringQualifier portQualifier) {
		this(originalStr, null, null, portQualifier, new EmbeddedAddress());
		embeddedAddress.addressProvider = valueProvider;
	}

	ParsedHost(String originalStr, int separatorIndices[], boolean normalizedFlags[], ParsedHostIdentifierStringQualifier labelsQualifier) {
		this(originalStr, separatorIndices, normalizedFlags, labelsQualifier, null);
	}
	
	ParsedHost(String originalStr, int separatorIndices[], boolean normalizedFlags[], ParsedHostIdentifierStringQualifier labelsQualifier, EmbeddedAddress embeddedAddress) {
		this.labelsQualifier = labelsQualifier;
		this.normalizedFlags = normalizedFlags;
		this.separatorIndices = separatorIndices;
		this.originalStr = originalStr;
		this.embeddedAddress = embeddedAddress == null ? NO_EMBEDDED_ADDRESS : embeddedAddress;
	}
	
	static class EmbeddedAddress implements Serializable {
		
		private static final long serialVersionUID = 4L;
		
		boolean isUNCIPv6Literal;
		boolean isReverseDNS;
		
		AddressStringException addressStringException;
		
		IPAddressProvider addressProvider;
	}
	
	public boolean isIPv6Address() {
		return hasEmbeddedAddress() && getAddressProvider().isProvidingIPv6();
	}
	
	public Integer getPort() {
		return labelsQualifier.getPort();
	}
	
	public String getService() {
		String serv = service;
		if(serv == null) {	
			CharSequence sv = labelsQualifier.getService();
			if(sv != null) {
				service = serv = sv.toString();
			}
		}
		return serv;
	}
	
	public Integer getNetworkPrefixLength() {
		return labelsQualifier.getNetworkPrefixLength();
	}
	
	public Integer getEquivalentPrefixLength() {
		return labelsQualifier.getEquivalentPrefixLength();
	}
	
	public IPAddress getMask() {
		return labelsQualifier.getMask();
	}
	
	public IPAddressProvider getAddressProvider() {
		return embeddedAddress.addressProvider;
	}
	
	private boolean hasEmbeddedAddress() {
		return embeddedAddress.addressProvider != null;
	}
	
	public boolean isAddressString() {
		return getAddressProvider() != null;
	}
	
	public IPAddress asAddress(IPVersion version) {
		if(hasEmbeddedAddress()) {
			return getAddressProvider().getProviderAddress(version);
		}
		return null;
	}
	
	public IPAddress asAddress() {
		if(hasEmbeddedAddress()) {
			return getAddressProvider().getProviderAddress();
		}
		return null;
	}
	
	public IPAddressString asGenericAddressString() {
		if(hasEmbeddedAddress()) {
			IPAddressProvider addressProvider = getAddressProvider();
			if(addressProvider.isProvidingAllAddresses()) {
				return new IPAddressString(IPAddress.SEGMENT_WILDCARD_STR, addressProvider.getParameters());
			} else if(addressProvider.isProvidingPrefixOnly()) {
				return new IPAddressString(IPAddressNetwork.getPrefixString(addressProvider.getProviderNetworkPrefixLength()), addressProvider.getParameters());
			} else if(addressProvider.isProvidingEmpty()) {
				return new IPAddressString("", addressProvider.getParameters());
			} else {
				IPAddress addr = addressProvider.getProviderAddress();
				return addr.toAddressString();
			}
		}
		return null;
	}

	public String[] getNormalizedLabels() {
		String labels[] = normalizedLabels;
		if(labels == null) {
			synchronized(this) {
				labels = normalizedLabels;
				if(labels == null) {
					if(hasEmbeddedAddress()) {
						IPAddressProvider addressProvider = getAddressProvider();
						IPAddress addr = addressProvider.getProviderAddress();
						if(addr == null) {
							if(addressProvider.isProvidingEmpty()) {
								return new String[0];
							}
							return new String[] {asGenericAddressString().toString()};
						}
						IPAddressSection section = addr.getSection();
						labels = section.getSegmentStrings();
					} else {
						labels = new String[separatorIndices.length];
						for(int i = 0, lastSep = -1; i < labels.length; i++) {
							int index = separatorIndices[i];
							if(normalizedFlags != null && !normalizedFlags[i]) {
								StringBuilder second = new StringBuilder((index - lastSep) - 1);
								for(int j = lastSep + 1; j < index; j++) {
									char c = originalStr.charAt(j);
									second.append((c >= 'A' && c <= 'Z') ? (char) (c + ('a' - 'A')) : c);
								}
								labels[i] = second.toString();
							} else {
								labels[i] = originalStr.substring(lastSep + 1, index);
							}
							lastSep = index;
						}
						separatorIndices = null;
						normalizedFlags = null;
					}
					normalizedLabels = labels;
				}
			}
		}
		return labels;
	}
	
	public String getHost() {
		String str = host;
		if(str == null) {
			if(originalStr.length() > 0) {
				synchronized(this) {
					str = host;
					if(str == null) {
						if(hasEmbeddedAddress()) {
							IPAddressProvider addressProvider = getAddressProvider();
							IPAddress addr = addressProvider.getProviderAddress();
							if(addr == null) {
								//note that this means prefix only (/16 or /64) is a valid host
								return asGenericAddressString().toString();
							}
							//port was stripped out 
							//mask and prefix removed by toNormalizedWildcardString
							//getSection() removes zone
							return addr.getSection().toCanonicalWildcardString();
						} else {
							StringBuilder builder = new StringBuilder(originalStr.length());
							String labels[] = getNormalizedLabels();
							builder.append(labels[0]);
							for(int i = 1; i < labels.length; i++) {
								builder.append(HostName.LABEL_SEPARATOR).append(labels[i]);
							}
							str = builder.toString();
						}
					}
				}
			} else {
				str = originalStr;
			}
			host = str;
		}
		return str;
	}
	
	public AddressStringException getAddressStringException() {
		return embeddedAddress.addressStringException;
	}
	
	public boolean isUNCIPv6Literal() {
		return embeddedAddress.isUNCIPv6Literal;
	}
	
	public boolean isReverseDNS() {
		return embeddedAddress.isReverseDNS;
	}
}

File: ipaddress-go/ipaddr/parsedhost.go
package ipaddr

import (
	"strings"
	//	"sync/atomic"
	//	"unsafe"
	"sync/atomic"
	"unsafe"
)

type embeddedAddress struct {
	isUNCIPv6Literal, isReverseDNS bool

	addressStringError AddressStringError

	addressProvider IPAddressProvider
}

var (
	// used by hosts
	//NO_EMBEDDED_ADDRESS *embeddedAddress                     = &embeddedAddress{}
	noQualifier *parsedHostIdentifierStringQualifier = &parsedHostIdentifierStringQualifier{}
)

type hostStrings struct {
	normalizedLabels []string
	host             string
}

type parsedHostCache struct {
	*hostStrings
}

type parsedHost struct {
	separatorIndices []int // can be nil
	normalizedFlags  []bool

	labelsQualifier parsedHostIdentifierStringQualifier

	embeddedAddress embeddedAddress

	originalStr string

	*parsedHostCache
}

func (host *parsedHost) getQualifier() *parsedHostIdentifierStringQualifier {
	return &host.labelsQualifier
}

func (host *parsedHost) isIPv6Address() bool {
	return host.hasEmbeddedAddress() && host.getAddressProvider().isProvidingIPv6()
}

func (host *parsedHost) getPort() Port {
	return host.labelsQualifier.getPort()
}

func (host *parsedHost) getService() string {
	return host.labelsQualifier.getService()
}

func (host *parsedHost) getNetworkPrefixLength() PrefixLen {
	return host.labelsQualifier.getNetworkPrefixLength()
}

func (host *parsedHost) getEquivalentPrefixLength() PrefixLen {
	return host.labelsQualifier.getEquivalentPrefixLength()
}

func (host *parsedHost) getMask() *IPAddress {
	return host.labelsQualifier.getMaskLower()
}

func (host *parsedHost) getAddressProvider() IPAddressProvider {
	return host.embeddedAddress.addressProvider
}

func (host *parsedHost) hasEmbeddedAddress() bool {
	return host.embeddedAddress.addressProvider != nil
}

func (host *parsedHost) isAddressString() bool {
	return host.getAddressProvider() != nil
}

func (host *parsedHost) asAddress() (*IPAddress, IncompatibleAddressError) {
	if host.hasEmbeddedAddress() {
		return host.getAddressProvider().getProviderAddress()
	}
	return nil, nil
}

func (host *parsedHost) mapString(addressProvider IPAddressProvider) string {
	if addressProvider.isProvidingAllAddresses() {
		return SegmentWildcardStr
		//} else if addressProvider.isProvidingPrefixOnly() {
		//return IPAddressNetwork.getPrefixString(addressProvider.getProviderNetworkPrefixLength())
	} else if addressProvider.isProvidingEmpty() {
		return ""
	}
	return host.originalStr
}

func (host *parsedHost) asGenericAddressString() *IPAddressString {
	if host.hasEmbeddedAddress() {
		addressProvider := host.getAddressProvider()
		if addressProvider.isProvidingAllAddresses() {
			return NewIPAddressStringParams(SegmentWildcardStr, addressProvider.getParameters())
			//} else if(addressProvider.isProvidingPrefixOnly()) {
			//	return new IPAddressString(IPAddressNetwork.getPrefixString(addressProvider.getProviderNetworkPrefixLength()), addressProvider.getParameters());
		} else if addressProvider.isProvidingEmpty() {
			return NewIPAddressStringParams("", addressProvider.getParameters())
		} else {
			//try {
			addr, err := addressProvider.getProviderAddress()
			if err != nil {
				return NewIPAddressStringParams(host.originalStr, addressProvider.getParameters())
			}
			return addr.ToAddressString()
			//} catch(IncompatibleAddressError e) {
			//return new IPAddressString(originalStr, addressProvider.getParameters());
			//}
		}
	}
	return nil
}

func (host *parsedHost) getHost() string {
	return host.buildStrings().host
}

func (host *parsedHost) buildStrings() *hostStrings {
	res := host.hostStrings
	if res == nil {
		var normalizedLabels []string
		var hostStr string
		if host.hasEmbeddedAddress() {
			addressProvider := host.getAddressProvider()
			addr, err := addressProvider.getProviderAddress()
			if err == nil {
				section := addr.GetSection()
				normalizedLabels = section.GetSegmentStrings()
				//port was stripped out
				//mask and prefix removed by toNormalizedWildcardString
				//getSection() removes zone
				hostStr = section.ToCanonicalWildcardString()
			} else {
				hostStr = host.mapString(addressProvider)
				if addressProvider.isProvidingEmpty() {
					normalizedLabels = []string{}
				} else {
					normalizedLabels = []string{hostStr}
				}
			}
		} else {
			normalizedLabels = make([]string, len(host.separatorIndices))
			normalizedFlags := host.normalizedFlags
			var first strings.Builder
			for i, lastSep := 0, -1; i < len(normalizedLabels); i++ {
				index := host.separatorIndices[i]
				if len(normalizedFlags) > 0 && !normalizedFlags[i] {
					var second strings.Builder
					second.Grow((index - lastSep) - 1)
					for j := lastSep + 1; j < index; j++ {
						c := host.originalStr[j]
						if c >= 'A' && c <= 'Z' {
							c = c + ('a' - 'A')
						}
						second.WriteByte(c)
					}
					normalizedLabels[i] = second.String()
				} else {
					normalizedLabels[i] = host.originalStr[lastSep+1 : index]
				}
				if i > 0 {
					first.WriteByte(LabelSeparator)
				}
				first.WriteString(normalizedLabels[i])
				lastSep = index
			}
			hostStr = first.String()
		}
		dataLoc := (*unsafe.Pointer)(unsafe.Pointer(&host.hostStrings))
		res = &hostStrings{
			normalizedLabels: normalizedLabels,
			host:             hostStr,
		}
		atomic.StorePointer(dataLoc, unsafe.Pointer(res))
	}
	return res
}

func (host *parsedHost) getNormalizedLabels() []string {
	return host.buildStrings().normalizedLabels
}

/*
	public AddressStringError getAddressStringException() { // this is an exception when something looks like reverse dns string or ip literal string and it is off a bit
		return embeddedAddress.addressStringError;
	}
*/

func (host *parsedHost) isUNCIPv6Literal() bool {
	return host.embeddedAddress.isUNCIPv6Literal
}

func (host *parsedHost) isReverseDNS() bool {
	return host.embeddedAddress.isReverseDNS
}
